### PR TITLE
[swiftc/msvc] Compiling with MSVC library

### PR DESCRIPTION
**Pull request by tinysun212**
#### What's in this pull request?

Some modifications for the ms-extension option of the clang.exe in the Visual Studio 2015 development environment

This patch is only for swiftc.exe. For the standard library or build script, other PR's will be used in later.
I used the library set of Visual Studio 2015 Update 1 and recent version of swift-clang as the compiler. If you are using the real MSVC compiler, more patch might be required.
#### Related bug number: ([SR-34](https://bugs.swift.org/browse/SR-34)) Port Swift to Windows

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

File: include/swift/Basic/Malloc.h [hunk 1]
```diff
@@ -19,12 +19,21 @@
...
 #include <cassert>
+#if defined(_MSC_VER)
+#include <malloc.h>
+#else
 #include <cstdlib>
+#endif
 
 namespace swift {
 
 // FIXME: Use C11 aligned_alloc or Windows _aligned_malloc if available.
```
Comment (jckarter) [inline include/swift/Basic/Malloc.h]: You can trim down this fixme a bit now.

File: include/swift/Basic/JSONSerialization.h [hunk 2]
```diff
@@ -481,6 +481,16 @@ struct ScalarTraits<uint32_t> {
   static bool mustQuote(StringRef) { return false; }
 };
 
+#if defined(_MSC_VER)
+// In MSVC, 'unsigned long' is 32bit size and different from uint32_t,
+// and it is used to define swift::sys::ProcessId.
```
Comment (jckarter) [inline include/swift/Basic/JSONSerialization.h]: It might be better to change `ProcessId` to use a `stdint` typedef.

File: lib/IDE/CodeCompletion.cpp [hunk 3]
```diff
@@ -4232,9 +4232,18 @@ void CodeCompletionCallbacksImpl::addKeywords(CodeCompletionResultSink &Sink) {
...
 
+#if defined(_MSC_VER)
+class CodeCompletionTypeContextAnalyzer;
+#endif
+
 namespace  {
   class ExprParentFinder : public ASTWalker {
+#if defined(_MSC_VER)
+    // MSVC needed '::' (http://reviews.llvm.org/D4443)
+    friend class ::CodeCompletionTypeContextAnalyzer;
```
Comment (jckarter) [inline lib/IDE/CodeCompletion.cpp]: Does it work portably if you say `friend class ::swift::CodeCompletionTypeContextAnalyzer;` ?

File: lib/ClangImporter/ClangImporter.cpp [hunk 4]
```diff
@@ -2039,7 +2039,9 @@ auto ClangImporter::Implementation::importFullName(
         if (!shouldImportAsInitializer(method, initPrefixLength,
                                        result.InitKind)) {
           // We cannot import this as an initializer anyway.
-          return { };
+          // Please don't use 'return { }' instead of this. MSVC compiler
+          // reports 'implicit initialization of field ErrorInfo'
```
Comment (gribozavr) [inline lib/ClangImporter/ClangImporter.cpp]: Please remove the comment, we don't leave such things in the source.

**Comment by gribozavr** [comment]
> Some modifications for the ms-compatibility mode of the clang-cl.exe in the Visual Studio 2015 development environment

Why is that interesting to support?  If you can use clang, why not use clang in its native mode?

File: lib/IDE/CodeCompletion.cpp (hunk 3 above)
Comment (tinysun212) [inline lib/IDE/CodeCompletion.cpp]: It reports ```error : no class named 'CodeCompletionTypeContextAnalyzer' in namespace 'swift'```.
CodeCompletion.cpp is not surrounded by ```namespace swift {```  ```}```.

File: include/swift/Basic/JSONSerialization.h (hunk 2 above)
Comment (tinysun212) [inline include/swift/Basic/JSONSerialization.h]: ProcessId in swift defined as ```typedef llvm::sys::ProcessInfo::ProcessId ProcessId;```.
LLVM ProcessInfo coded.
```
struct ProcessInfo {
#if defined(LLVM_ON_UNIX)
  typedef pid_t ProcessId;
#elif defined(LLVM_ON_WIN32)
  typedef unsigned long ProcessId; // Must match the type of DWORD on Windows.
  .....
```
Changing ProcessId in swift means changing or ignoring LLVM definition.
I think reusing the LLVM definition is better.

File: lib/IDE/CodeCompletion.cpp (hunk 3 above)
Comment (jckarter) [inline lib/IDE/CodeCompletion.cpp]: It should probably be inside the anonymous namespace, then.

**Comment by tinysun212** [comment]
For build and link with MS C runtime library, including their headers are necessary. Without ms-compatibility mode, they are cannot be compiled.

File: include/swift/Basic/JSONSerialization.h (hunk 2 above)
Comment (jckarter) [inline include/swift/Basic/JSONSerialization.h]: Yeah, you're right. Understandable that LLVM wouldn't want to include all of windows.h just for `DWORD`.

**Comment by jckarter** [comment]
Could we use native Clang mode for swiftc, but MS compatibility mode for the runtime?

**Comment by jckarter** [comment]
Or could we avoid needing to parse Visual C++'s libraries by using libc++? Do pure C `windows.h`-level headers really need MSVC compatibility?

**Comment by tinysun212** [comment]
I'm confused on options in comments.
I enabled two options -fms-extension, -fms-compatibility. The former required for understanding MS includes. I'll check if any issue occurs when the latter option is removed.

**Comment by jckarter** [comment]
Sounds good. AIUI `-fms-extension` should be fine, since that enables purely additive MSVC features, whereas `-fms-compatibility` causes Clang to deliberately emulate MSVC bugs. We'd prefer not to have to deal with the latter.

**Comment by tinysun212** [comment]
After remove -fms-compatibility option, some files can be removed from PR. Also, I modified some comments and rebranched.

File: lib/Immediate/Immediate.cpp [hunk 5]
```diff
@@ -37,9 +37,12 @@
 #include "llvm/Transforms/IPO.h"
 #include "llvm/Transforms/IPO/PassManagerBuilder.h"
 #include "llvm/Support/Path.h"
-
+#if defined(_MSC_VER)
+#include "Windows.h"
+#define dlopen(x, y)  LoadLibrary(x)
```
Comment (jckarter) [inline lib/Immediate/Immediate.cpp]: Instead of a macro, can we use a wrapper function with a conditional implementation?

File: lib/ClangImporter/ClangImporter.cpp [hunk 6]
```diff
@@ -2039,7 +2039,7 @@ auto ClangImporter::Implementation::importFullName(
         if (!shouldImportAsInitializer(method, initPrefixLength,
                                        result.InitKind)) {
           // We cannot import this as an initializer anyway.
-          return { };
+          return ImportedName();
```
Comment (jckarter) [inline lib/ClangImporter/ClangImporter.cpp]: Is this change still necessary?

File: lib/Sema/CSDiag.cpp [hunk 7]
```diff
@@ -845,6 +845,15 @@ namespace {
...
     void filterList(ArrayRef<CallArgParam> actualArgs);
+#if _MSC_VER
+    // MSVC confused when choosing a correct method.
+    void filterList(SmallVector<CallArgParam, 4> actualArgs) {
+      return filterList(ArrayRef<CallArgParam>(actualArgs));
+    }
+    void filterList(CallArgParam actualArgs) {
+      return filterList(ArrayRef<CallArgParam>(actualArgs));
+    }
+#endif
```
Comment (jckarter) [inline lib/Sema/CSDiag.cpp]: Is this change necessary without `-fms-compatibility`?

File: lib/SILOptimizer/Mandatory/ConstantPropagation.cpp [hunk 8]
```diff
@@ -875,7 +875,14 @@ processFunction(SILFunction &F, bool EnableDiagnostics,
...
         WorkList.remove(I);
         I->eraseFromParent();
+#if defined(_MSC_VER)
+      // The default parameter didn't work in MSVC.
+      },
+      []() {},
+      []() {});
+#else
       });
+#endif
```
Comment (jckarter) [inline lib/SILOptimizer/Mandatory/ConstantPropagation.cpp]: Is this change necessary without `-fms-compatibility`?

File: lib/ClangImporter/ClangImporter.cpp (hunk 6 above)
Comment (tinysun212) [inline lib/ClangImporter/ClangImporter.cpp]: Yes. It is probably related to the template library.

File: lib/Sema/CSDiag.cpp (hunk 7 above)
Comment (tinysun212) [inline lib/Sema/CSDiag.cpp]: Yes, it is.

File: lib/SILOptimizer/Mandatory/ConstantPropagation.cpp (hunk 8 above)
Comment (tinysun212) [inline lib/SILOptimizer/Mandatory/ConstantPropagation.cpp]: Yes, it also couldn't be removed.

File: lib/Sema/CSDiag.cpp (hunk 7 above)
Comment (lattner) [inline lib/Sema/CSDiag.cpp]: Please don't add #ifdefs for this.  If there is a problem resolving the overload, just rename the methods to be distinct, and update the callers.

**Comment by tinysun212** [comment]
After removing ```-fms-compatablitity-version=19```, ```return { };``` in ```ClangImporter.cpp``` is revoked. I think the option is not ignored even if ```-fno-ms-compatibility``` was used.

But the patch which the ```[](){}``` are required in ```ConstantPropagation.cpp``` could not revoked by removing the option. I think it is related to the ```<functional>``` header implemention.

Two or more default parameter of functional in constructor was not resolved. Removing ifdefs in ```ConstantPropagation.cpp```, I changed one constructor of three default parameters in ```Local.h``` to two constructors of one default parameter.

**Comment by tinysun212** [comment]
For the patch of ```ConstantPropagation.cpp```, I'll check more if it can be revoked.

**Comment by tinysun212** [comment]
The error of ```return { };```  in ```ClangImporter.cpp``` was not for the bug emulation in ms-compatibility. I think it was the difference between c++11 and c++14. I wrote following simple code and tested it.

```
tt.cpp
------

enum class MyType { SampleValue };

class Opt {
public:
	explicit Opt() {}
};

struct Astr {
	MyType  myt = MyType::SampleValue;
	Opt  opt;
};

Astr Foo2() {
	return{};
}
```

```
$ clang++ -c  tt.cpp  -std=c++11   -->  NO ERROR
```

```
$ clang++ -c  tt.cpp  -std=c++14
tt.cpp:15:9: error: chosen constructor is explicit in copy-initialization
        return{};
               ^
tt.cpp:6:11: note: constructor declared here
        explicit Opt() {}
                 ^
tt.cpp:11:7: note: in implicit initialization of field 'opt' with omitted initializer
        Opt  opt;
             ^
1 error generated.
```

By verifying ```clang -v```, when the ```-fms-compatablitity-version=19``` option is used, clang driver inserted ```-std=c++14``` option to the fronted.

Anyway, we will not use the compatibility option, no problem.

**Comment by tinysun212** [comment]
For the patch of ```ConstantPropagation.cpp```, it can be revoked, but has a little problem.

I have two clang version for Windows.
One is the clang.exe I built from swift-clang repository, and I made this PR with it.
Other is the clang 3.7 with MS CodeGen which released by MS last year. 

MS clang and MS cl compiler recognize well two or more functional default parameter in constructor.
But my clang was not recognize it. I think MS clang has some different code not yet applied to swift-clang.

I cannot use MS clang for ```ConstantPropagation.cpp``` now, it generates a fatal error at other position instead. I want to keep this patch.

**Comment by gottesmm** [comment]
@tinysun212 Can you smoosh together the patches?

**Comment by gribozavr** [comment]
Do we really need to make any changes to the compiler?  Why can't we use `clang.exe` to build the compiler, and `clang-cl.exe` to build the runtime?

**Comment by tinysun212** [comment]
@gribozavr  I explained the reason to need this patch and to use ```-fms-extensions``` option. I think your question is about to using the option.

I want to run the compiler (```swiftc.exe```) under Windows. We already have the Cygwin ported compiler, but it uses rarely used library packages. In this time, I'm building ```swiftc.exe``` using MS standard C++ library - more widely used runtime.

The compiler source is written in C++ and acceptable by clang in OS X and Linux.
And the compiler source uses the standard library such as ```<functional>, <map>, <string>, ...```.
Libraries have the corresponding header files and those header files must be compiled to build ```swiftc.exe```.

The clang package does not including its own C/C++ standard library.

In Linux or Cygwin, unless installing the LLVM libc++, the clang package installs and uses the libraries built for gcc.

In Windows, we have to choose the C/C++ standard library to build with.

Last month, I choosed the gcc library for Cygwin port. It works.

In this time, I choosed the MS standard C++ library. This is widely used standard library option in Windows.

MS header files corresponding to the standard C++ library use some ms-extension features and not acceptable by plain clang. On the other hand, the header files written for gcc in other OS are well compiled by clang.

So, to running the compiler in widely used ways under Windows, it couldn't to be compiled with plain clang.

**Comment by stefanstefan2001** [comment]
Ping?

**Comment by tinysun212** [comment]
@jxyang has more changes than me in his repository.

**Comment by tinysun212** [comment]
@gribozavr, The difference between ```clang.exe``` and ```clang-cl.exe``` is the driver options for clang frontend. Anyway, this PR is suitable to ```clang.exe``` with the option ```-fms-extension``` and not use```'-fms-compatibility```. In ```clang-cl.exe```, both of them are enabled.

To build the compiler ```swift.exe``` on Windows (MSVC, not Cygwin), we should use the MS standard C++ library and headers. We can not drop the option ```-fms-extension``` because the library can't be used without the option.

Unfortunately, the option ```-fms-extension``` caused some code changes, but this is the minimum patch to build the compiler on Windows.

**Comment by tinysun212** [comment]
ping?

**Comment by CodaFi** [comment]
@tinysun212 It looks like you have merge conflicts.  Could you please rebase onto master.

File: lib/Basic/TaskQueue.cpp [hunk 9]
```diff
@@ -23,7 +23,7 @@ using namespace swift;
 using namespace swift::sys;
 
 // Include the correct TaskQueue implementation.
-#if LLVM_ON_UNIX && !defined(__CYGWIN__)
+#if LLVM_ON_UNIX && !defined(__CYGWIN__) && !defined(_MSC_VER)
```
Comment (jrose-apple) [inline lib/Basic/TaskQueue.cpp]: Why is LLVM_ON_UNIX even set for MSVC?

File: lib/Driver/Compilation.cpp [hunk 10]
```diff
@@ -631,8 +631,15 @@ int Compilation::performSingleCommand(const Job *Cmd) {
   const char *ExecPath = Cmd->getExecutable();
   const char **argv = Argv.data();
 
-  for (auto &envPair : Cmd->getExtraEnvironment())
+  for (auto &envPair : Cmd->getExtraEnvironment()) {
+#if defined(_MSC_VER)
+    std::string envStr = envPair.first;
```
Comment (jrose-apple) [inline lib/Driver/Compilation.cpp]: Nitpick: please use an llvm::SmallString instead of std::string, so that smaller settings don't require an extra heap allocation.

**Comment by tinysun212** [comment]
@CodaFi, @jrose-apple, thanks for your reviews, it is rebased and changed.

File: include/swift/SILOptimizer/Utils/Local.h [hunk 11]
```diff
@@ -485,14 +485,18 @@ class CastOptimizer {
 
 public:
   CastOptimizer(std::function<void (SILInstruction *I, ValueBase *V)> ReplaceInstUsesAction,
```
Comment (jrose-apple) [inline include/swift/SILOptimizer/Utils/Local.h]: Can I ask why this change was necessary?

**Comment by jrose-apple** [comment]
Looks reasonable to me. Thanks, @tinysun212!

@swift-ci Please test

File: include/swift/SILOptimizer/Utils/Local.h (hunk 11 above)
Comment (tinysun212) [inline include/swift/SILOptimizer/Utils/Local.h]: Using the clang with the ms-extension option (not ms-compatibility option), the line of the constructor having two default value of dummy lambda expression is compiled, but three default value constructor is not compiled.

Comment (jrose-apple) [inline include/swift/SILOptimizer/Utils/Local.h]: That's very strange. Mind leaving a comment before I merge?

**Comment by tinysun212** [comment]
@jrose-apple, I applied your reviews.

**Comment by jrose-apple** [comment]
Thanks, merged!

**Comment by tinysun212** [comment]
Thanks, this merging I have long awaited. ^^
